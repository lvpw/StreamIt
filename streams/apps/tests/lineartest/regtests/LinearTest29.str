/** 
 * Test out doing upsampling/downsampling with FIR filter
 * to show that the new 
 **/
void->void pipeline LinearTest29 {
  add FloatSource();
  add Expander(2);
  add MockFIRFilter(5,2);
  add Compressor(3);
  add FloatSink();
}

void->float stateful filter FloatSource {
  float x;
  init {
    x = 0;
  }
  work push 1 {
    push(x);
    x = x + 1;
  }
}

float->void filter FloatSink {
  work pop 1 {
    println(pop());
  }
}

/** This is a mock up of an FIR filter. N is where the counting starts from **/
float->float filter MockFIRFilter(int size, int N) {
  float[size] weights;
  init {
    for (int i=0; i<size; i++) {
      weights[i] = N + i;
    }
  }

  work pop 1 peek size push 1 {
    float sum = 0;
    /** do the MAC like normal **/
    for (int i=0; i<size; i++) {
      sum = sum + (peek(i) * weights[i]);
    }
    push(sum);
    pop();
  }
}

/** Expander (add L-1 zeros for every 1 input) **/
float->float filter Expander(int L) {
  work pop 1 peek 1 push L {
    push(pop());
    for (int i=0; i<(L-1); i++) {
      push(0);
    }
  }
}

/** Compressor (drop every M-1 values for every one pushed **/
float->float filter Compressor(int M) {
  work pop M peek M push 1 {
    push(pop());
    for (int i=0; i<(M-1); i++) {
      pop();
    }
  }
}
