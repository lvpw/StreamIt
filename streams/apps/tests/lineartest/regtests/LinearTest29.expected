Compressor_1.matrix
[[0.0+0.0i]
 [0.0+0.0i]
 [1.0+0.0i]]
Compressor_1.vector
[[0.0+0.0i]]
Expander_1.matrix
[[0.0+0.0i 1.0+0.0i]]
Expander_1.vector
[[0.0+0.0i 0.0+0.0i]]
LinearWrapper_1.matrix
[[0.0+0.0i 0.0+0.0i]
 [5.0+0.0i 0.0+0.0i]
 [3.0+0.0i 6.0+0.0i]
 [0.0+0.0i 4.0+0.0i]
 [0.0+0.0i 2.0+0.0i]]
LinearWrapper_1.vector
[[0.0+0.0i 0.0+0.0i]]
MockFIRFilter_1.matrix
[[6.0+0.0i]
 [5.0+0.0i]
 [4.0+0.0i]
 [3.0+0.0i]
 [2.0+0.0i]]
MockFIRFilter_1.vector
[[0.0+0.0i]]
